lib/libmd/sha.3
.\"
.\" ----------------------------------------------------------------------------
.\" "THE BEER-WARE LICENSE" (Revision 42):
.\" <phk@FreeBSD.org> wrote this file.  As long as you retain this notice you
.\" can do whatever you want with this stuff. If we meet some day, and you think
.\" this stuff is worth it, you can buy me a beer in return.   Poul-Henning Kamp
.\" ----------------------------------------------------------------------------
.\"
.\" 	From: Id: mdX.3,v 1.14 1999/02/11 20:31:49 wollman Exp
.\"
.Dd February 6, 2023
.Dt SHA 3
.Os
.Sh NAME
.Nm SHA_Init ,
.Nm SHA_Update ,
.Nm SHA_Final ,
.Nm SHA_End ,
.Nm SHA_File ,
.Nm SHA_FileChunk ,
.Nm SHA_Data ,
.Nm SHA1_Init ,
.Nm SHA1_Update ,
.Nm SHA1_Final ,
.Nm SHA1_End ,
.Nm SHA1_File ,
.Nm SHA1_FileChunk ,
.Nm SHA1_Data
.Nd calculate the FIPS 160 and 160-1 ``SHA'' message digests
.Sh LIBRARY
.Lb libmd
.Sh SYNOPSIS
.In sys/types.h
.In sha.h
.Ft void
.Fn SHA_Init "SHA_CTX *context"
.Ft void
.Fn SHA_Update "SHA_CTX *context" "const unsigned char *data" "size_t len"
.Ft void
.Fn SHA_Final "unsigned char digest[20]" "SHA_CTX *context"
.Ft "char *"
.Fn SHA_End "SHA_CTX *context" "char *buf"
.Ft "char *"
.Fn SHA_File "const char *filename" "char *buf"
.Ft "char *"
.Fn SHA_FileChunk "const char *filename" "char *buf" "off_t offset" "off_t length"
.Ft "char *"
.Fn SHA_Data "const unsigned char *data" "unsigned int len" "char *buf"
.Ft void
.Fn SHA1_Init "SHA_CTX *context"
.Ft void
.Fn SHA1_Update "SHA_CTX *context" "const unsigned char *data" "size_t len"
.Ft void
.Fn SHA1_Final "unsigned char digest[20]" "SHA_CTX *context"
.Ft "char *"
.Fn SHA1_End "SHA_CTX *context" "char *buf"
.Ft "char *"
.Fn SHA1_File "const char *filename" "char *buf"
.Ft "char *"
.Fn SHA1_FileChunk "const char *filename" "char *buf" "off_t offset" "off_t length"
.Ft "char *"
.Fn SHA1_Data "const unsigned char *data" "unsigned int len" "char *buf"
.Sh DESCRIPTION
The
.Li SHA_
and
.Li SHA1_
functions calculate a 160-bit cryptographic checksum (digest)
for any number of input bytes.
A cryptographic checksum is a one-way
hash function; that is, it is computationally impractical to find
the input corresponding to a particular output.
This net result is
a
.Dq fingerprint
of the input-data, which does not disclose the actual input.
.Pp
SHA (or SHA-0) is the original Secure Hash Algorithm specified in FIPS 160.
It was quickly proven insecure, and has been superseded by SHA-1.
SHA-0 is included for compatibility purposes only.
.Pp
The
.Fn SHA1_Init ,
.Fn SHA1_Update ,
and
.Fn SHA1_Final
functions are the core functions.
Allocate an
.Vt SHA_CTX ,
initialize it with
.Fn SHA1_Init ,
run over the data with
.Fn SHA1_Update ,
and finally extract the result using
.Fn SHA1_Final ,
which will also erase the
.Vt SHA_CTX .
.Pp
.Fn SHA1_End
is a wrapper for
.Fn SHA1_Final
which converts the return value to a 41-character
(including the terminating '\e0')
ASCII string which represents the 160 bits in hexadecimal.
.Pp
.Fn SHA1_File
calculates the digest of a file, and uses
.Fn SHA1_End
to return the result.
If the file cannot be opened, a null pointer is returned.
.Fn SHA1_FileChunk
is similar to
.Fn SHA1_File ,
but it only calculates the digest over a byte-range of the file specified,
starting at
.Fa offset
and spanning
.Fa length
bytes.
If the
.Fa length
parameter is specified as 0, or more than the length of the remaining part
of the file,
.Fn SHA1_FileChunk
calculates the digest from
.Fa offset
to the end of file.
.Fn SHA1_Data
calculates the digest of a chunk of data in memory, and uses
.Fn SHA1_End
to return the result.
.Pp
When using
.Fn SHA1_End ,
.Fn SHA1_File ,
or
.Fn SHA1_Data ,
the
.Fa buf
argument can be a null pointer, in which case the returned string
is allocated with
.Xr malloc 3
and subsequently must be explicitly deallocated using
.Xr free 3
after use.
If the
.Fa buf
argument is non-null it must point to at least 41 characters of buffer space.
.Sh ERRORS
The
.Fn SHA1_End
function called with a null buf argument may fail and return NULL if:
.Bl -tag -width Er
.It Bq Er ENOMEM
Insufficient storage space is available.
.El
.Pp
The
.Fn SHA1_File
and
.Fn SHA1_FileChunk
may return NULL when underlying
.Xr open 2 ,
.Xr fstat 2 ,
.Xr lseek 2 ,
or
.Xr SHA1_End 2
fail.
.Sh SEE ALSO
.Xr md4 3 ,
.Xr md5 3 ,
.Xr ripemd 3 ,
.Xr sha256 3 ,
.Xr sha512 3 ,
.Xr skein 3
.Sh HISTORY
These functions appeared in
.Fx 4.0 .
.Sh AUTHORS
The core hash routines were implemented by Eric Young based on the
published
FIPS standards.
.Sh BUGS
The SHA1 algorithm has been proven to be vulnerable to practical collision
attacks and should not be relied upon to produce unique outputs,
.Em nor should it be used as part of a new cryptographic signature scheme.


NEWS
Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader
Jun 16, 2011: Add tar:compat-2x option to emulate broken libarchive 2.x
   handling of pax UTF-8 headers
Apr 25, 2011: Refactor read_open() into a collection of single-item setters;
   support the old interfaces as wrappers
Apr 12, 2011: Split disk writer into separate POSIX and Windows implementations
Apr 10, 2011: Improvements to character translations on Windows.
Mar 30, 2011: More work to return errors instead of calling abort()

lib/hcrypto/sha.c
/*
 * Copyright (c) 1995 - 2001 Kungliga Tekniska Högskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#include "hash.h"
#include "sha.h"

#define A m->counter[0]
#define B m->counter[1]
#define C m->counter[2]
#define D m->counter[3]
#define E m->counter[4]
#define X data

void
SHA1_Init (struct sha *m)
{
  m->sz[0] = 0;
  m->sz[1] = 0;
  A = 0x67452301;
  B = 0xefcdab89;
  C = 0x98badcfe;
  D = 0x10325476;
  E = 0xc3d2e1f0;
}


#define F0(x,y,z) CRAYFIX((x & y) | (~x & z))
#define F1(x,y,z) (x ^ y ^ z)
#define F2(x,y,z) ((x & y) | (x & z) | (y & z))
#define F3(x,y,z) F1(x,y,z)

#define K0 0x5a827999
#define K1 0x6ed9eba1
#define K2 0x8f1bbcdc
#define K3 0xca62c1d6

#define DO(t,f,k) \
do { \
  uint32_t temp; \
 \
  temp = cshift(AA, 5) + f(BB,CC,DD) + EE + data[t] + k; \
  EE = DD; \
  DD = CC; \
  CC = cshift(BB, 30); \
  BB = AA; \
  AA = temp; \
} while(0)

static inline void
calc (struct sha *m, uint32_t *in)
{
  uint32_t AA, BB, CC, DD, EE;
  uint32_t data[80];
  int i;

  AA = A;
  BB = B;
  CC = C;
  DD = D;
  EE = E;

  for (i = 0; i < 16; ++i)
    data[i] = in[i];
  for (i = 16; i < 80; ++i)
    data[i] = cshift(data[i-3] ^ data[i-8] ^ data[i-14] ^ data[i-16], 1);

  /* t=[0,19] */

  DO(0,F0,K0);
  DO(1,F0,K0);
  DO(2,F0,K0);
  DO(3,F0,K0);
  DO(4,F0,K0);
  DO(5,F0,K0);
  DO(6,F0,K0);
  DO(7,F0,K0);
  DO(8,F0,K0);
  DO(9,F0,K0);
  DO(10,F0,K0);
  DO(11,F0,K0);
  DO(12,F0,K0);
  DO(13,F0,K0);
  DO(14,F0,K0);
  DO(15,F0,K0);
  DO(16,F0,K0);
  DO(17,F0,K0);
  DO(18,F0,K0);
  DO(19,F0,K0);

  /* t=[20,39] */

  DO(20,F1,K1);
  DO(21,F1,K1);
  DO(22,F1,K1);
  DO(23,F1,K1);
  DO(24,F1,K1);
  DO(25,F1,K1);
  DO(26,F1,K1);
  DO(27,F1,K1);
  DO(28,F1,K1);
  DO(29,F1,K1);
  DO(30,F1,K1);
  DO(31,F1,K1);
  DO(32,F1,K1);
  DO(33,F1,K1);
  DO(34,F1,K1);
  DO(35,F1,K1);
  DO(36,F1,K1);
  DO(37,F1,K1);
  DO(38,F1,K1);
  DO(39,F1,K1);

  /* t=[40,59] */

  DO(40,F2,K2);
  DO(41,F2,K2);
  DO(42,F2,K2);
  DO(43,F2,K2);
  DO(44,F2,K2);
  DO(45,F2,K2);
  DO(46,F2,K2);
  DO(47,F2,K2);
  DO(48,F2,K2);
  DO(49,F2,K2);
  DO(50,F2,K2);
  DO(51,F2,K2);
  DO(52,F2,K2);
  DO(53,F2,K2);
  DO(54,F2,K2);
  DO(55,F2,K2);
  DO(56,F2,K2);
  DO(57,F2,K2);
  DO(58,F2,K2);
  DO(59,F2,K2);

  /* t=[60,79] */

  DO(60,F3,K3);
  DO(61,F3,K3);
  DO(62,F3,K3);
  DO(63,F3,K3);
  DO(64,F3,K3);
  DO(65,F3,K3);
  DO(66,F3,K3);
  DO(67,F3,K3);
  DO(68,F3,K3);
  DO(69,F3,K3);
  DO(70,F3,K3);
  DO(71,F3,K3);
  DO(72,F3,K3);
  DO(73,F3,K3);
  DO(74,F3,K3);
  DO(75,F3,K3);
  DO(76,F3,K3);
  DO(77,F3,K3);
  DO(78,F3,K3);
  DO(79,F3,K3);

  A += AA;
  B += BB;
  C += CC;
  D += DD;
  E += EE;
}

/*
 * From `Performance analysis of MD5' by Joseph D. Touch <touch@isi.edu>
 */

#if !defined(WORDS_BIGENDIAN) || defined(_CRAY)
static inline uint32_t
swap_uint32_t (uint32_t t)
{
#define ROL(x,n) ((x)<<(n))|((x)>>(32-(n)))
  uint32_t temp1, temp2;

  temp1   = cshift(t, 16);
  temp2   = temp1 >> 8;
  temp1  &= 0x00ff00ff;
  temp2  &= 0x00ff00ff;
  temp1 <<= 8;
  return temp1 | temp2;
}
#endif

struct x32{
  unsigned int a:32;
  unsigned int b:32;
};

void
SHA1_Update (struct sha *m, const void *v, size_t len)
{
  const unsigned char *p = v;
  size_t old_sz = m->sz[0];
  size_t offset;

  m->sz[0] += len * 8;
  if (m->sz[0] < old_sz)
      ++m->sz[1];
  offset = (old_sz / 8)  % 64;
  while(len > 0){
    size_t l = min(len, 64 - offset);
    memcpy(m->save + offset, p, l);
    offset += l;
    p += l;
    len -= l;
    if(offset == 64){
#if !defined(WORDS_BIGENDIAN) || defined(_CRAY)
      int i;
      uint32_t SHA1current[16];
      struct x32 *us = (struct x32*)m->save;
      for(i = 0; i < 8; i++){
	SHA1current[2*i+0] = swap_uint32_t(us[i].a);
	SHA1current[2*i+1] = swap_uint32_t(us[i].b);
      }
      calc(m, SHA1current);
#else
      calc(m, (uint32_t*)m->save);
#endif
      offset = 0;
    }
  }
}

void
SHA1_Final (void *res, struct sha *m)
{
  unsigned char zeros[72];
  unsigned offset = (m->sz[0] / 8) % 64;
  unsigned int dstart = (120 - offset - 1) % 64 + 1;

  *zeros = 0x80;
  memset (zeros + 1, 0, sizeof(zeros) - 1);
  zeros[dstart+7] = (m->sz[0] >> 0) & 0xff;
  zeros[dstart+6] = (m->sz[0] >> 8) & 0xff;
  zeros[dstart+5] = (m->sz[0] >> 16) & 0xff;
  zeros[dstart+4] = (m->sz[0] >> 24) & 0xff;
  zeros[dstart+3] = (m->sz[1] >> 0) & 0xff;
  zeros[dstart+2] = (m->sz[1] >> 8) & 0xff;
  zeros[dstart+1] = (m->sz[1] >> 16) & 0xff;
  zeros[dstart+0] = (m->sz[1] >> 24) & 0xff;
  SHA1_Update (m, zeros, dstart + 8);
  {
      int i;
      unsigned char *r = (unsigned char*)res;

      for (i = 0; i < 5; ++i) {
	  r[4*i+3] = m->counter[i] & 0xFF;
	  r[4*i+2] = (m->counter[i] >> 8) & 0xFF;
	  r[4*i+1] = (m->counter[i] >> 16) & 0xFF;
	  r[4*i]   = (m->counter[i] >> 24) & 0xFF;
      }
  }
#if 0
  {
    int i;
    uint32_t *r = (uint32_t *)res;

    for (i = 0; i < 5; ++i)
      r[i] = swap_uint32_t (m->counter[i]);
  }
#endif
}
